tasks/service/async/src/test/java/org/projectnessie/nessie/tasks/async/BaseTasksAsync.java

```diff
@@ -0,0 +1,236 @@
+        stages.stream()
+            .map(CompletionStage::toCompletableFuture)
+            .map(
+                cf -> {
+                  try {
+                    return cf.get();
+                  } catch (Exception e) {
+                    throw new RuntimeException(e);
+                  }
+                })
+            .collect(Collectors.toSet());
+
+    soft.assertThat(stages)
+        .hasSize(num)
+        .extracting(CompletionStage::toCompletableFuture)
+        .allMatch(CompletableFuture::isDone);
+
+    soft.assertThat(nums)
+        .containsExactlyInAnyOrderElementsOf(
+            IntStream.range(0, num).boxed().collect(Collectors.toSet()));
+  }
+
+  @Test
+  public void callExceptionallyWorks() throws Exception {
+    TasksAsync async = tasksAsync();
+
+    CompletionStage<Void> stage =
+        async.call(
+            () -> {
+              throw new RuntimeException("hello");
+            });
+
+    Throwable mappedFailure =
+        stage.handle((result, failure) -> failure).toCompletableFuture().get();
+
+    soft.assertThat(mappedFailure)
+        .isInstanceOf(CompletionException.class)
+        .extracting(Throwable::getCause)
+        .isInstanceOf(RuntimeException.class)
+        .extracting(Throwable::getMessage)
+        .isEqualTo("hello");
+  }
+
+  @Test
+  public void scheduledExceptionallyWorks() throws Exception {
+    TasksAsync async = tasksAsync();
+
+    CompletionStage<Void> stage =
+        async
+            .schedule(
+                () -> {
+                  throw new RuntimeException("hello");
+                },
+                Instant.now())
+            .toCompletionStage();
+
+    Throwable mappedFailure =
+        stage.handle((result, failure) -> failure).toCompletableFuture().get();
+
+    soft.assertThat(mappedFailure)
+        .isInstanceOf(CompletionException.class)
+        .extracting(Throwable::getCause)
+        .isInstanceOf(RuntimeException.class)
+        .extracting(Throwable::getMessage)
+        .isEqualTo("hello");
+  }
+
+  @ParameterizedTest
+  @ValueSource(ints = {1, 10, 1000})
+  public void scheduleWorks(int num) throws InterruptedException {
+    TasksAsync async = tasksAsync();
+
+    Semaphore sem = new Semaphore(0);
+
+    for (int i = 0; i < num; i++) {
+      async.schedule(sem::release, async.clock().instant());
+    }
+
+    soft.assertThat(sem.tryAcquire(num, 10, TimeUnit.SECONDS)).isTrue();
+  }
+
+  @Test
+  public void cancelDoesNotError() {
+    // See below
+    TasksAsync async = tasksAsync();
+    ScheduledHandle handle =
+        async.schedule(() -> {}, async.clock().instant().plus(1500, ChronoUnit.MILLIS));
+    soft.assertThatCode(handle::cancel).doesNotThrowAnyException();
+  }
+
+  @Test
+  @Disabled(
+      "Testing scheduled/timer cancellation requires either tweaking the inner workings of Vert.X / ScheduledExecutorService or results in extremely long test durations.")
+  public void cancelReallyWorks() throws InterruptedException {
```

Comment:
I think it's important to test cancellation, and besides, Vertx has weird cancellation effects.

For the Java pool impl, you could use the below test:

```java
  public void cancelReallyWorks() throws InterruptedException {
    TasksAsync async = tasksAsync();

    CountDownLatch started = new CountDownLatch(1);
    CountDownLatch finished = new CountDownLatch(1);

    CompletionStage<Void> handle =
        async.schedule(
            () -> {
              try {
                started.countDown();
                new Semaphore(0).acquire();
              } catch (InterruptedException ignored) {
                finished.countDown();
              }
            },
            async.clock().instant());
    soft.assertThat(handle).isNotNull();

    assertThat(started.await(10, TimeUnit.SECONDS)).isTrue();
    handle.toCompletableFuture().cancel(true);
    assertThat(finished.await(10, TimeUnit.SECONDS)).isTrue();
  }
```

(The example above includes my suggestion of getting rid of `ScheduleHandle`.)

For the Vrtx impl, it's tricky because Vertx, being event-loop based, does _not_ interrupt threads when the timer is cancelled. So indeed it's very hard to test whether cancellation works because, in practice, timer tasks _cannot_ be cancelled if they were started already. Cancelling the timer only has an effect if the timer wasn't triggered yet.

As a side note: it's important to note that for the Vertx impl, runnables scheduled with `TasksAsync.schedule()` _must not block_.